data_pipeline/europepmc/europepmc_pipeline.py

```diff
@@ -0,0 +1,194 @@
+        search_context
+    )
+    response = requests.get(url, params=params)
+    return get_valid_json_from_response(response)
+
+
+def get_filtered_article_data(data: dict, fields_to_return: Optional[Sequence[str]]) -> dict:
+    if not fields_to_return:
+        return data
+    return {key: value for key, value in data.items() if key in fields_to_return}
+
+
+def iter_article_data(
+    source_config: EuropePmcSourceConfig,
+    search_context: EuropePmcSearchContext
+) -> Iterable[dict]:
+    LOGGER.info('source_config: %r', source_config)
+    response_json = get_article_response_json_from_api(
+        source_config,
+        search_context
+    )
+    return (
+        get_filtered_article_data(data, source_config.fields_to_return)
+        for data in iter_article_data_from_response_json(response_json)
+    )
+
+
+def save_state_to_s3_for_config(
+    state_config: EuropePmcStateConfig,
+    start_date_str: str
+):
+    upload_s3_object(
+        bucket=state_config.state_file.bucket_name,
+        object_key=state_config.state_file.object_name,
+        data_object=start_date_str
+    )
+
+
+def get_next_start_date_str_for_end_date_str(
+    end_date_str: str
+):
+    end_date = date.fromisoformat(end_date_str)
+    next_start_date = end_date + timedelta(days=1)
+    return next_start_date.isoformat()
+
+
+def get_search_context_for_start_date_str(
+    start_date_str: str,
+    max_days: Optional[int] = None
+) -> EuropePmcSearchContext:
+    today = date.today()
+    yesterday = today - timedelta(days=1)
+    if not max_days:
+        end_date = yesterday
+    else:
+        end_date = date.fromisoformat(start_date_str) + timedelta(days=max_days - 1)
+        if end_date >= today:
+            end_date = yesterday
+    end_date_str = end_date.isoformat()
+    return EuropePmcSearchContext(
+        start_date_str=start_date_str,
+        end_date_str=end_date_str
+    )
+
+
+def load_state_from_s3_for_config(
+    state_config: EuropePmcStateConfig
+) -> str:
+    try:
+        return download_s3_object_as_string_or_file_not_found_error(
+            bucket=state_config.state_file.bucket_name,
+            object_key=state_config.state_file.object_name
+        )
+    except FileNotFoundError:
+        LOGGER.info('state file not found, returning initial state')
+        return state_config.initial_state.start_date_str
+
+
+def fetch_article_data_from_europepmc_and_load_into_bigquery(
+    config: EuropePmcConfig
+):
+    start_date_str = load_state_from_s3_for_config(
+        config.state
+    )
+    search_context = get_search_context_for_start_date_str(
+        start_date_str,
+        max_days=config.source.max_days
+    )
+    LOGGER.debug('search_context: %r', search_context)
+    if search_context.is_empty_period():
+        LOGGER.info('empty period, skip processing')
+        return
+    batch_size = config.batch_size
+    data_iterable = iter_article_data(
+        config.source,
+        search_context
+    )
```

Comment:
Note: Before loading data into BQ it would be ideal to add `imported_timestamp` in.

Reply:
I've added provenance with an `imported_timestamp` as well as other potential useful information.